external/fv3fit/fv3fit/emulation/layers/fields.py

```diff
@@ -0,0 +1,192 @@
+            sample_out.shape[-1], name=f"unscaled_{self.name}"
+        )
+
+        if denormalize is not None:
+            self.denorm = get_denorm_class(denormalize)(
+                name=f"denormalized_{self.name}"
+            )
+            self.denorm.fit(sample_out)
+        else:
+            self.denorm = tf.keras.layers.Lambda(lambda x: x)
+
+        self.relu = tf.keras.layers.ReLU()
+        self.use_relu = enforce_positive
+
+    def call(self, tensor):
+
+        tensor = self.unscaled(tensor)
+        tensor = self.denorm(tensor)
+
+        if self.use_relu:
+            tensor = self.relu(tensor)
+
+        return tensor
+
+
+class IncrementStateLayer(tf.keras.layers.Layer):
+    """
+    Layer for incrementing states with a tendency tensor
+
+    Attributes:
+        dt_sec: timestep delta in seconds
+    """
+
+    def __init__(self, dt_sec: int, *args, dtype=tf.float32, **kwargs):
+
+        self.dt_sec = tf.constant(dt_sec, dtype=dtype)
+        super().__init__(*args, **kwargs)
+
+    def call(self, initial: tf.Tensor, tendency: tf.Tensor) -> tf.Tensor:
+        """
+        Increment state with tendency * timestep
+
+        args:
+            tensors: Input state field and corresponding tendency tensor to
+                increment by.
+        """
+        return initial + tendency * self.dt_sec
+
+
+class IncrementedFieldOutput(tf.keras.layers.Layer):
+    """
+    Add input tensor to an output tensor using timestep increment.
+    This residual-style architecture is analogous to learning tendency-based updates.
+    """
+
+    def __init__(
+        self,
+        sample_out: tf.Tensor,
+        dt_sec: int,
+        *args,
+        denormalize: Optional[str] = None,
+        enforce_positive: bool = False,
+        **kwargs,
+    ):
+        """
+        Args:
+            sample_out: Output sample for variable to set shape
+                and fit denormalization layer.
+            dt_sec: Timestep length in seconds to use for incrementing
+                input state.
+            denormalize: denormalize layer key to use on
+                the dense layer output
+        """
+        super().__init__(*args, **kwargs)
+
+        self.tendency = FieldOutput(
+            sample_out,
+            denormalize=denormalize,
+            enforce_positive=False,
+            name=f"residual_{self.name}",
+        )
+        self.increment = IncrementStateLayer(dt_sec, name=f"increment_{self.name}")
+        self.use_relu = enforce_positive
+        self.relu = tf.keras.layers.ReLU()
+
+    def call(self, field_input, network_output):
+
+        tendency = self.tendency(network_output)
+        tensor = self.increment(field_input, tendency)
+
+        if self.use_relu:
+            tensor = self.relu(tensor)
+
+        return tensor
+
+    def get_tendency_output(self, network_output):
+        return self.tendency(network_output)
+
+
+class CombineInputs(tf.keras.layers.Layer):
```

Comment:
Seeing as how this is instantiated by a check on the architecture type, I think this class would be more at home in `layers.architecture` than `layers.fields`.